Windows Support

KingRial opened the pull request:
Since I loved the library, I just added a quick support for windows based on [WMI](https://docs.microsoft.com/en-us/windows/win32/wmisdk/querying-wmi) queries.

I am pretty new to GoLang so, this is far from being perfect or fully completed, but at least it's a start.

Hope it could be of any help.

jaypipes commented:
Hi! Thank you so much @KingRial for this PR! Sorry for the delay in reviewing, was on PTO... will review in next day or so and test out on a Windows VM. Thanks! :)

KingRial commented:
Hello @jaypipes!
It is my pleasure to collaborate with this project.

A quick overview of what I am offering:

- A support for **windows HW inspection**
I still know there are big area of improvements but, at least, there are the main informations and anyone can improve them
- **Memory struct**
added how many phsysical "memory banks" are present and their main informations
(let me know if you prefer a different way to handle them)

As an inspection library I presumed it could be interesting in having also some non-HW data.
**N.B.** If it distorts the library philosophy I will remove them!

- **NIC struct**
started to add some SW configuration: IP / Gateways / DHCP
- **Packages**
Added the ability to list the installed packages/softwares on the HW (right now it supports windows only but adding the linux/darwin support is just a matter of time)

Let me know if  can help you somehow; but please remember I am pretty new to GoLang ^^

jaypipes commented:
Hello again, @KingRial! Just checking in with you. I'm writing this from a Windows machine where I am testing out your patch and doing a full code review. I'm also reviewing the github.com/StackExchange/wmi library at the same time to evaluate including it as a dependency for ghw.

Would it be possible to separate your pull request into two individual PRs, the first being the changes to add Windows support without any additional fields to the core `ghw` API structs (such as the new `MemoryInfo.MemoryBanks` field), the second being any changes to the core `ghw` structs?

jaypipes commented on baseboard_windows.go:
```diff
@@ -0,0 +1,32 @@
+func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
+	// Getting disk drives from WMI
+	var win32BaseboardDescriptions []Win32_BaseBoard
+	q1 := wmi.CreateQuery(&win32BaseboardDescriptions, "")
```
You don't need to use the `wmi.CreateQuery()` function here. Instead, simply create a const string with the WQL query, like so:

```go
const (
        wqlBaseboard = "SELECT Manufacturer, SerialNumber, Tag, Version FROM Win32_BaseBoard"
)

type win32Baseboard struct {
	Manufacturer string
	SerialNumber string
	Tag          string
	Version      string
}

func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
        var win32BaseboardDescriptions []win32Baseboard
        if err := wmi.Query(wqlBaseboard, &win32BaseboardDescriptions); err != nil {
		return err
	}
	if len(win32BaseboardDescriptions) > 0 {
		info.AssetTag = win32BaseboardDescriptions[0].Manufacturer
		info.SerialNumber = win32BaseboardDescriptions[0].Manufacturer
		info.Vendor = win32BaseboardDescriptions[0].Manufacturer
		info.Version = win32BaseboardDescriptions[0].Manufacturer
	}
	return nil
}
```

KingRial replied:
That is true.

The main reason I was using the same pattern was because I could just change the **win32Baseboard** struct obtaining dynamically the correct updated query.

Pretty usefull if you are lazy like me and you are always exploring WMI tables to find the correct information ^^

Let me know if you still think it's better to use a static query instead of having a dynamic one built from the structure.

jaypipes replied:
Yeah, I think I'd prefer to just use a static const string query versus using an exported/public struct named the same as the WQL database table. It's just cleaner IMHO :)

jaypipes commented on bios_windows.go:
```diff
@@ -0,0 +1,29 @@
+
+import "github.com/StackExchange/wmi"
+
+type CIM_BIOSElement struct {
```
Same comment about naming of the struct. Please use `biosElement` and just pass a const string of `SELECT InstallDate, Manufacturer, Version FROM CMI_BIOSElement` to `cmi.Query()` below.

jaypipes commented on bios_windows.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+func (ctx *context) biosFillInfo(info *BIOSInfo) error {
+	// Getting disk drives from WMI
```
just remove the above comment :)

KingRial replied:
Done ^^

jaypipes commented on block.go:
```diff
@@ -108,13 +108,14 @@ type Disk struct {
+	Label       string `json:"label"`
+	MountPoint  string `json:"mount_point"`
+	SizeBytes   uint64 `json:"size_bytes"`
+	UsableBytes uint64 `json:"usable_bytes"`
```
Please pull this change into a separate pull request :)

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+		readOnly = false
+	}
+	return readOnly
+}
```
The above function can be simplified to just this:

```go
func toReadOnly(access uint16) bool {
	// See Access property from: https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-diskpartition
        return access == 0x1
}
```

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+				}
+			}
+		}
+		// Appending Disk
```
you can remove the above comment :)

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+							//fmt.Printf("--------------- Disk drive %#v\n", diskdrive)
+							//fmt.Printf("--------------- Disk Partition %#v\n", diskpartition)
+							//fmt.Printf("--------------- Logical Disk %#v\n", logicaldisk)
+							// Appending Partition
```
you can remove the above comment

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+						if logicaldisktodiskpartition.Antecedent == desiredAntecedent && logicaldisktodiskpartition.Dependent == desiredDependent {
+							//fmt.Printf("--------------- Disk drive %#v\n", diskdrive)
+							//fmt.Printf("--------------- Disk Partition %#v\n", diskpartition)
+							//fmt.Printf("--------------- Logical Disk %#v\n", logicaldisk)
```
Please remove above commented-out prints.

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+						var desiredAntecedent = "\\\\" + diskpartition.SystemName + "\\root\\cimv2:" + diskpartition.CreationClassName + ".DeviceID=\"" + diskpartition.DeviceID + "\""
+						var desiredDependent = "\\\\" + logicaldisk.SystemName + "\\root\\cimv2:" + logicaldisk.CreationClassName + ".DeviceID=\"" + logicaldisk.DeviceID + "\""
+						//fmt.Printf("-- Antecedent\n%#v\n%#v\n", desiredAntecedent, logicaldisktodiskpartition.Antecedent)
+						//fmt.Printf("-- Dependent\n%#v\n%#v\n", desiredDependent, logicaldisktodiskpartition.Dependent)
```
Please remove above commented-out prints.

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+					for _, logicaldisktodiskpartition := range win32LogicalDiskToPartitionDescriptions {
+						//fmt.Printf("\nRelation %#v\n", logicaldisktodiskpartition)
+						//fmt.Printf("logicaldisk %#v\n", logicaldisk)
+						//fmt.Printf("diskpartition %#v\n", diskpartition)
```
Please remove above commented-out prints.

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+			// Finding disk partition linked to current disk drive
+			if diskdrive.Index == diskpartition.DiskIndex {
+				disk.PhysicalBlockSizeBytes = diskpartition.BlockSize
+				//fmt.Printf("Disk Partition %#v\n", diskdrive)
```
Please remove above commented-out prints.

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+			StorageController:      toStorageController(diskdrive.InterfaceType),
+			BusType:                toBusType(diskdrive.InterfaceType),
+			BusPath:                UNKNOWN, // TODO: add information
+			//NUMANodeID:             node,
```
Please remove above. NUMANodeID isn't a field of Disk :)

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+	q4 := wmi.CreateQuery(&win32LogicalDiskDescriptions, "")
+	if err := wmi.Query(q4, &win32LogicalDiskDescriptions); err != nil {
+		return err
+	}
```
for each of the above, please create a separate helper function that returns the slice of appropriate structs, where the helper function uses a const string `cmi.Query()` instead of the `cmi.CreateQuery()` function.

Example for the disk drives:

```go
const (
        wqlDiskDrives = `
SELECT Caption, CreationClassName, Description, DeviceID,
    Index, InterfaceType. Manufacturer, MediaType, Model, Name,
    SerialNumber, Size, TotalCylinders, TotalHeads, TotalSectors,
    TotalTracks, TracksPerCylinder
FROM Win32_DiskDrive`
)

func (ctx *context) blockFillInfo(info *BlockInfo) error {
        win32DiskDriveDescriptions, err := getDiskDrives()
        if err != nil {
                return err
        }
...
}

func getDiskDrives() ([]win32DiskDrive, error) {
	// Getting disk drives from WMI
	var win32DiskDriveDescriptions []win32DiskDrive
	if err := wmi.Query(wqlDiskDrives, &win32DiskDriveDescriptions); err != nil {
		return nil, err
	}
        return win3232DiskDriveDescriptions, nil
}
```

jaypipes commented on block_windows.go:
```diff
@@ -0,0 +1,218 @@
+}
+
+/*type Win32_DiskDriveToDiskPartition struct {
+}*/
```
unused above. please remove.

jaypipes commented on chassis_windows.go:
```diff
@@ -0,0 +1,40 @@
+	"github.com/StackExchange/wmi"
+)
+
+type CIM_Chassis struct {
```
Same comment as above about using an unexported struct name and a const string for the WQL query.

jaypipes commented on cpu_windows.go:
```diff
@@ -0,0 +1,62 @@
+	for index, description := range win32descriptions {
+		p := &Processor{
+			Id:         index,            // TODO: how to get a decent "Physical ID" to use ?
+			Model:      description.Name, // description.ProcessorID description.Manufacturer description.Name
```
please remove comment at end of this line.

jaypipes commented on cpu_windows.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_Processor struct {
```
same comment about the struct name and using a const query string

jaypipes commented on gpu_windows.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_VideoController struct {
```
same comment about the struct name and using a const query string

jaypipes commented on gpu_windows.go:
```diff
@@ -0,0 +1,52 @@
+		cards = append(cards, card)
+	}
+	//car.DeviceInfo
+	//card.Node
```
Please remove the above comments

jaypipes commented on memory.go:
```diff
@@ -11,9 +11,19 @@ import (
+	SerialNumber string `json:"vendor"`
+	SizeBytes    int64  `json:"size_bytes"`
+	Vendor       string `json:"vendor"`
+}
```
Please separate these changes into a different pull request.

jaypipes commented on memory_windows.go:
```diff
@@ -0,0 +1,76 @@
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_OperatingSystem struct {
```
same comment about the struct name and using a const query string

jaypipes commented on net.go:
```diff
@@ -10,17 +10,26 @@ import (
+	Gateway     string `json:"gateway"`
+	IPv4        string `json:"ipv4"`
+	IPv6        string `json:"ipv6"`
+}
```
Please move the changes adding this new NICConfiguration to a separate pull request where we can discuss this feature in isolation from the Windows-specific patch here.

jaypipes commented on net_windows.go:
```diff
@@ -0,0 +1,141 @@
+	return nil
+}
+
+type Win32_NetworkAdapter struct {
```
same comment about the struct name and using a const query string

jaypipes commented on packages.go:
```diff
@@ -0,0 +1,68 @@
+//
```
Please move all the additional packages module to a new pull request. It is not related to the Windows PR or the NIC Configuration or memory banks work.

jaypipes commented on product_windows.go:
```diff
@@ -0,0 +1,42 @@
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_ComputerSystemProduct struct {
```
same comment about the struct name and using a const query string

jaypipes commented on pci_windows.go:
```diff
@@ -0,0 +1,87 @@
+			Name: UNKNOWN,
+		},
+	}
+	// Getting disk drives from WMI
```
Not getting disk drives :)

jaypipes commented on pci_windows.go:
```diff
@@ -0,0 +1,87 @@
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
```
Let's leave PCI information on Windows to a separate pull request, OK? I need to do some more research on just how relevant the PnPEntity thing in WMI is to a PCI device. Please just pull all the PCI code out into a separate pull request.

KingRial replied:
Right now GPU e PCI are linked to try to follow the same pattern used on linux.

I had the same doubts, but the PNPDeviceID was the only working way I could get the Vendor/Product device data about the GPU.

If you prefer I can split GPU and PCI leaving the PCI to posterity

jaypipes replied:
Yeah, let's split the GPU and PCI stuff for a separate PR where we can discuss this topic in isolation.

jaypipes commented:
@KingRial can you email me at jaypipes at gmail dot com? Wanted to check with you about something privately... thanks!

KingRial commented on baseboard_windows.go:
```diff
@@ -0,0 +1,32 @@
+}
+
+func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
+	// Getting disk drives from WMI
```
My bad!

KingRial commented on baseboard_windows.go:
```diff
@@ -0,0 +1,32 @@
+
+import "github.com/StackExchange/wmi"
+
+type Win32_BaseBoard struct {
```
I agree on changing to unexported struct

KingRial commented:
Closing, since current request is going to be splitted in different pull requests
